README.md
mLabs <img src="https://avatars1.githubusercontent.com/u/29802441?s=200&v=4" alt="drawing" width="45px"/> - Teste back-end

## API de controle de estacionamento (v1.0)

[![Build Status](https://drone.geomk.com.br/api/badges/bispo/parking/status.svg)](https://drone.geomk.com.br/bispo/parking) ![standard-readme compliant](https://img.shields.io/badge/ruby-2.6.5-brightgreen.svg?style=flat-square) ![standard-readme compliant](https://img.shields.io/badge/rails-6-brightgreen.svg?style=flat-square)

## Objetivo

Testar os conhecimentos, código e a organização do candidato no desenvolvimento de uma API usando linguagem de programação ruby.

## Conteúdo do Repositório

1. [Este README](README.md).
2. Histórico de alterações e releases do sistema: [CHANGELOG.md](CHANGELOG.md).
3. [Documentação do desenvolvimento](HOWTO.md).
4. [Fontes do Projeto](/source/mlabs-parking)

## Sumário

- [Visão Geral](#vis%c3%a3o-geral)
- [Principais Funcionalidades](#principais-funcionalidades)
- [Especificações Técnicas](#especifica%c3%a7%c3%b5es-t%c3%a9cnicas)
  - [**Entrada no estacionamento**](#entrada-no-estacionamento)
  - [**Pagamento do estacionamento**](#pagamento-do-estacionamento)
  - [**Saida do estacionamento**](#saida-do-estacionamento)
- [Instalação](#instala%c3%a7%c3%a3o)
- [Configuração](#configura%c3%a7%c3%a3o)
- [Uso no desenvolvimento](#uso-no-desenvolvimento)
- [Mantenedores](#mantenedores)
- [License MIT](#license-mit)


# Visão Geral

API para controle de estacionamento

# Principais Funcionalidades

- Entra no estacionamento
- Pagamento do estacionamento
- Saída de estacionamento
- Histórico de uma dada placa

# Especificações Técnicas

**Routes**

|Prefix      |Verb  |URI Pattern              |Controller#Action  |
|------------|:----:|-------------------------|-------------------|
|            |POST  |/parking                 |parking#create     |
|parking_pay |GET   |/parking/:parking_id/pay |parking#pay        |
|parking_out |GET   |/parking/:parking_id/out |parking#out        |
|            |GET   |/parking/:plate          |parking#historic   |


**Entrada no estacionamento**
----
>  Criação do ticket de estacionamento no momento da entrada

* **URL**

  /parking

* **Method:**
  
  `POST`
  
*  **No URL Params**

* **Data Params**

  { plate: "XYZ-1234" }

* **Success Response:**
  
  * **Code:** 200
    **Content:** `{ id:1, plate: "XYZ-1234"}`


**Pagamento do estacionamento**
----
>  Liberação da saida por meio do pagamento do ticket

* **URL**

  /parking/:parking_id/pay

* **Method:**
  
  `GET`
  
* **No URL Params**
    
* **No Data Params**

* **Success Response:**
  
  * **Code:** 200
    **Content:** `{ id:1, plate: "XYZ-1234", paid:true }`
 
* **Error Response:**

  * **Code:** 404 NOT FOUND
    **Content:** `{ message : "Couldn't find Parking" }`

**Saida do estacionamento**
----

>  Saida do estacionamento

* **URL**

  /parking/:parking_id/out

* **Method:**
  
  `GET`
  
* **No URL Params**

* **No Data Params**

* **Success Response:**
  
  * **Code:** 200
    **Content:** `{ id:1, plate: "XYZ-1234", time: "15 minutes", paid:true }`
 
* **Error Response:**

  * **Code:** 402 PAYMENT REQUIRED
    **Content:** `{ message : "payment required" }`


# Instalação

# Configuração

> Environment
>
> - RAILS_ENV = development | test | production
> 

# Uso no desenvolvimento

> Levantar a aplicação em modo desenvolvimento

Clone o projeto
```
$ git clone https://github.com/bispogeomk/mlabs-parking.git
$ cd mlabs-parking
```

Rode a aplcação usando docker-compose

```bash
$ docker-compose up
```


# Mantenedores
 - Moacir Bispo - <bispo@det.ufc.br>

# License MIT



CHANGELOG.md
# Changelog (Parking API)

Todas as mudanças notáveis ​​neste projeto serão documentadas neste arquivo.

O formato é baseado em [Manter um Changelog] (https://keepachangelog.com/pt-BR/1.0.0/)
e este projeto adere a [Semantic Versioning] (https://semver.org/spec/v2.0.0.html).

> Tipos de mudanças:
>
> - Adicionado
> - Modificado
> - Obsoleto
> - Removido
> - Corrigido
> - Segurança

## [0.0.1] - 2019-19-19
### Adicionado
- este CHANGELOG
- README.md
- Estrutura básica do projeto

> Lembretes:
>
> - Changelogs são para humanos, não máquinas.
> - Deve haver uma entrada para cada versão.
> - Alterações do mesmo tipo devem ser agrupadas.
> - Versões e seções devem ser vinculáveis (com links).
> - A versão mais recente vem em primeiro lugar.
> - A data de lançamento de cada versão é exibida.


source/mlabs-parking/spec/rails_helper.rb
# This file is copied to spec/ when you run 'rails generate rspec:install'
require 'spec_helper'
ENV['RAILS_ENV'] ||= 'test'

require File.expand_path('../config/environment', __dir__)

# Prevent database truncation if the environment is production
abort("The Rails environment is running in production mode!") if Rails.env.production?
require 'rspec/rails'
# Add additional requires below this line. Rails is not loaded until this point!

# require database cleaner at the top level
require 'database_cleaner'

# Requires supporting ruby files with custom matchers and macros, etc, in
# spec/support/ and its subdirectories. Files matching `spec/**/*_spec.rb` are
# run as spec files by default. This means that files in spec/support that end
# in _spec.rb will both be required and run as specs, causing the specs to be
# run twice. It is recommended that you do not name files matching this glob to
# end with _spec.rb. You can configure this pattern with the --pattern
# option on the command line or in ~/.rspec, .rspec or `.rspec-local`.
#
# The following line is provided for convenience purposes. It has the downside
# of increasing the boot-up time by auto-requiring all files in the support
# directory. Alternatively, in the individual `*_spec.rb` files, manually
# require only the support files necessary.
#
# Dir[Rails.root.join('spec', 'support', '**', '*.rb')].each { |f| require f }
Dir[Rails.root.join('spec/support/**/*.rb')].each { |f| require f }

# configure shoulda matchers to use rspec as the test framework and full matcher libraries for rails
Shoulda::Matchers.configure do |config|
  config.integrate do |with|
    with.test_framework :rspec
    with.library :rails
  end
end

# Checks for pending migrations and applies them before tests are run.
# If you are not using ActiveRecord, you can remove these lines.
begin
  ActiveRecord::Migration.maintain_test_schema!
rescue ActiveRecord::PendingMigrationError => e
  puts e.to_s.strip
  exit 1
end
RSpec.configure do |config|
  # Remove this line if you're not using ActiveRecord or ActiveRecord fixtures
  config.fixture_path = "#{::Rails.root}/spec/fixtures"

  # add `FactoryBot` methods
  config.include FactoryBot::Syntax::Methods

  config.include RequestSpecHelper, type: :request

  # start by truncating all the tables but then use the faster transaction strategy the rest of the time.
  config.before(:suite) do
    DatabaseCleaner.clean_with(:truncation)
    DatabaseCleaner.strategy = :transaction
  end

  # start the transaction strategy as examples are run
  config.around(:each) do |example|
    DatabaseCleaner.cleaning do
      example.run
    end
  end

  # If you're not using ActiveRecord, or you'd prefer not to run each of your
  # examples within a transaction, remove the following line or assign false
  # instead of true.
  config.use_transactional_fixtures = true

  # RSpec Rails can automatically mix in different behaviours to your tests
  # based on their file location, for example enabling you to call `get` and
  # `post` in specs under `spec/controllers`.
  #
  # You can disable this behaviour by removing the line below, and instead
  # explicitly tag your specs with their type, e.g.:
  #
  #     RSpec.describe UsersController, :type => :controller do
  #       # ...
  #     end
  #
  # The different available types are documented in the features, such as in
  # https://relishapp.com/rspec/rspec-rails/docs
  config.infer_spec_type_from_file_location!

  # Filter lines from Rails gems in backtraces.
  config.filter_rails_from_backtrace!
  # arbitrary gems may also be filtered via:
  # config.filter_gems_from_backtrace("gem name")
end


source/mlabs-parking/spec/factories/parkings.rb
FactoryBot.define do
  factory :parking do
    plate { "AAA-9876" }
    car_in { "2019-10-18 03:58:24" }
    car_out { nil }
    minutes { nil }
    left { false }
    paid { false }
  end
end


source/mlabs-parking/config/environments/test.rb
# The test environment is used exclusively to run your application's
# test suite. You never need to work with it otherwise. Remember that
# your test database is "scratch space" for the test suite and is wiped
# and recreated between test runs. Don't rely on the data there!

Rails.application.configure do
  config.hosts = nil
  # Settings specified here will take precedence over those in config/application.rb.
  
  config.cache_classes = false

  # Do not eager load code on boot. This avoids loading your whole application
  # just for the purpose of running a single test. If you are using a tool that
  # preloads Rails for running tests, you may have to set it to true.
  config.eager_load = false

  # Configure public file server for tests with Cache-Control for performance.
  config.public_file_server.enabled = true
  config.public_file_server.headers = {
    'Cache-Control' => "public, max-age=#{1.hour.to_i}"
  }

  # Show full error reports and disable caching.
  config.consider_all_requests_local       = true
  config.action_controller.perform_caching = false
  config.cache_store = :null_store

  # Raise exceptions instead of rendering exception templates.
  config.action_dispatch.show_exceptions = false

  # Disable request forgery protection in test environment.
  config.action_controller.allow_forgery_protection = false

  # Store uploaded files on the local file system in a temporary directory.
  config.active_storage.service = :test

  config.action_mailer.perform_caching = false

  # Tell Action Mailer not to deliver emails to the real world.
  # The :test delivery method accumulates sent emails in the
  # ActionMailer::Base.deliveries array.
  config.action_mailer.delivery_method = :test

  # Print deprecation notices to the stderr.
  config.active_support.deprecation = :stderr

  # Raises error for missing translations.
  # config.action_view.raise_on_missing_translations = true
end


source/mlabs-parking/db/schema.rb
# This file is auto-generated from the current state of the database. Instead
# of editing this file, please use the migrations feature of Active Record to
# incrementally modify your database, and then regenerate this schema definition.
#
# This file is the source Rails uses to define your schema when running `rails
# db:schema:load`. When creating a new database, `rails db:schema:load` tends to
# be faster and is potentially less error prone than running all of your
# migrations from scratch. Old migrations may fail to apply correctly if those
# migrations use external dependencies or application code.
#
# It's strongly recommended that you check this file into your version control system.

ActiveRecord::Schema.define(version: 2019_10_18_035319) do

  create_table "parkings", force: :cascade do |t|
    t.string "plate"
    t.datetime "car_in"
    t.datetime "car_out"
    t.integer "minutes"
    t.boolean "paid"
    t.boolean "left"
    t.datetime "created_at", precision: 6, null: false
    t.datetime "updated_at", precision: 6, null: false
  end

end


source/mlabs-parking/config/application.rb
require_relative 'boot'

require "rails"
# Pick the frameworks you want:
require "active_model/railtie"
require "active_job/railtie"
require "active_record/railtie"
require "active_storage/engine"
require "action_controller/railtie"
require "action_mailer/railtie"
require "action_mailbox/engine"
require "action_text/engine"
require "action_view/railtie"
require "action_cable/engine"
# require "sprockets/railtie"
# require "rails/test_unit/railtie"

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
Bundler.require(*Rails.groups)

module MlabsParking
  class Application < Rails::Application
    # Initialize configuration defaults for originally generated Rails version.
    config.load_defaults 6.0

    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration can go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded after loading
    # the framework and any gems in your application.

    # Only loads a smaller set of middleware suitable for API only apps.
    # Middleware like session, flash, cookies can be added back manually.
    # Skip views, helpers and assets when generating a new resource.
    config.api_only = true
  end
end


source/mlabs-parking/config/environments/production.rb
Rails.application.configure do
  # Settings specified here will take precedence over those in config/application.rb.

  # Code is not reloaded between requests.
  config.cache_classes = true

  # Eager load code on boot. This eager loads most of Rails and
  # your application in memory, allowing both threaded web servers
  # and those relying on copy on write to perform better.
  # Rake tasks automatically ignore this option for performance.
  config.eager_load = true

  # Full error reports are disabled and caching is turned on.
  config.consider_all_requests_local       = false

  # Ensures that a master key has been made available in either ENV["RAILS_MASTER_KEY"]
  # or in config/master.key. This key is used to decrypt credentials (and other encrypted files).
  # config.require_master_key = true

  # Disable serving static files from the `/public` folder by default since
  # Apache or NGINX already handles this.
  config.public_file_server.enabled = ENV['RAILS_SERVE_STATIC_FILES'].present?

  # Enable serving of images, stylesheets, and JavaScripts from an asset server.
  # config.action_controller.asset_host = 'http://assets.example.com'

  # Specifies the header that your server uses for sending files.
  # config.action_dispatch.x_sendfile_header = 'X-Sendfile' # for Apache
  # config.action_dispatch.x_sendfile_header = 'X-Accel-Redirect' # for NGINX

  # Store uploaded files on the local file system (see config/storage.yml for options).
  config.active_storage.service = :local

  # Mount Action Cable outside main process or domain.
  # config.action_cable.mount_path = nil
  # config.action_cable.url = 'wss://example.com/cable'
  # config.action_cable.allowed_request_origins = [ 'http://example.com', /http:\/\/example.*/ ]

  # Force all access to the app over SSL, use Strict-Transport-Security, and use secure cookies.
  # config.force_ssl = true

  # Use the lowest log level to ensure availability of diagnostic information
  # when problems arise.
  config.log_level = :debug

  # Prepend all log lines with the following tags.
  config.log_tags = [ :request_id ]

  # Use a different cache store in production.
  # config.cache_store = :mem_cache_store

  # Use a real queuing backend for Active Job (and separate queues per environment).
  # config.active_job.queue_adapter     = :resque
  # config.active_job.queue_name_prefix = "mlabs_parking_production"

  config.action_mailer.perform_caching = false

  # Ignore bad email addresses and do not raise email delivery errors.
  # Set this to true and configure the email server for immediate delivery to raise delivery errors.
  # config.action_mailer.raise_delivery_errors = false

  # Enable locale fallbacks for I18n (makes lookups for any locale fall back to
  # the I18n.default_locale when a translation cannot be found).
  config.i18n.fallbacks = true

  # Send deprecation notices to registered listeners.
  config.active_support.deprecation = :notify

  # Use default logging formatter so that PID and timestamp are not suppressed.
  config.log_formatter = ::Logger::Formatter.new

  # Use a different logger for distributed setups.
  # require 'syslog/logger'
  # config.logger = ActiveSupport::TaggedLogging.new(Syslog::Logger.new 'app-name')

  if ENV["RAILS_LOG_TO_STDOUT"].present?
    logger           = ActiveSupport::Logger.new(STDOUT)
    logger.formatter = config.log_formatter
    config.logger    = ActiveSupport::TaggedLogging.new(logger)
  end

  # Do not dump schema after migrations.
  config.active_record.dump_schema_after_migration = false

  # Inserts middleware to perform automatic connection switching.
  # The `database_selector` hash is used to pass options to the DatabaseSelector
  # middleware. The `delay` is used to determine how long to wait after a write
  # to send a subsequent read to the primary.
  #
  # The `database_resolver` class is used by the middleware to determine which
  # database is appropriate to use based on the time delay.
  #
  # The `database_resolver_context` class is used by the middleware to set
  # timestamps for the last write to the primary. The resolver uses the context
  # class timestamps to determine how long to wait before reading from the
  # replica.
  #
  # By default Rails will store a last write timestamp in the session. The
  # DatabaseSelector middleware is designed as such you can define your own
  # strategy for connection switching and pass that into the middleware through
  # these configuration options.
  # config.active_record.database_selector = { delay: 2.seconds }
  # config.active_record.database_resolver = ActiveRecord::Middleware::DatabaseSelector::Resolver
  # config.active_record.database_resolver_context = ActiveRecord::Middleware::DatabaseSelector::Resolver::Session
end


source/mlabs-parking/config/environments/development.rb
Rails.application.configure do
  config.hosts = nil
  # Settings specified here will take precedence over those in config/application.rb.

  # In the development environment your application's code is reloaded on
  # every request. This slows down response time but is perfect for development
  # since you don't have to restart the web server when you make code changes.
  config.cache_classes = false

  # Do not eager load code on boot.
  config.eager_load = false

  # Show full error reports.
  config.consider_all_requests_local = true

  # Enable/disable caching. By default caching is disabled.
  # Run rails dev:cache to toggle caching.
  if Rails.root.join('tmp', 'caching-dev.txt').exist?
    config.cache_store = :memory_store
    config.public_file_server.headers = {
      'Cache-Control' => "public, max-age=#{2.days.to_i}"
    }
  else
    config.action_controller.perform_caching = false

    config.cache_store = :null_store
  end

  # Store uploaded files on the local file system (see config/storage.yml for options).
  config.active_storage.service = :local

  # Don't care if the mailer can't send.
  config.action_mailer.raise_delivery_errors = false

  config.action_mailer.perform_caching = false

  # Print deprecation notices to the Rails logger.
  config.active_support.deprecation = :log

  # Raise an error on page load if there are pending migrations.
  config.active_record.migration_error = :page_load

  # Highlight code that triggered database queries in logs.
  config.active_record.verbose_query_logs = true


  # Raises error for missing translations.
  # config.action_view.raise_on_missing_translations = true

  # Use an evented file watcher to asynchronously detect changes in source code,
  # routes, locales, etc. This feature depends on the listen gem.
  config.file_watcher = ActiveSupport::EventedFileUpdateChecker
end


source/mlabs-parking/app/controllers/parking_controller.rb
class ParkingController < ApplicationController
    before_action :set_parking, only: [:update, :destroy]
    # before_action :set_parkings_by_plate, only: [:show]

    # GET /parking
    def index
        @parkings = Parking.all
        json_response(@parkings)
    end

    # POST /parkings
    def create
        @parking = Parking.create!(parking_params)
        @parking.car_in = Time.now
        @parking.car_out = nil
        @parking.minutes = nil
        @parking.paid = false
        @parking.left = false
        @parking.save
        json_response({
            "id" => @parking.id, 
            "plate" => @parking.plate}, 201)
    end

    # PUT /parking/:id
    def update
        @parking.update(parking_params)
        head :no_content
    end

    # DELETE /parking/:id
    def destroy
        @parking.destroy
        head :no_content
    end

    # GET /parking/:parking_id/pay
    def pay
        @parking = Parking.find(params[:parking_id])
        @parking.paid = true
        @parking.save
        json_response(@parking)
    end

    # GET /parking/:parking_id/out
    def out
        @parking = Parking.find(params[:parking_id])
        if @parking.paid
            @parking.car_out = Time.now
            @parking.left = true
            diff_time = @parking.car_out - @parking.car_in
            @parking.minutes = (diff_time/60).to_i
            @parking.save
            json_response({ 
                "id" => @parking.id,
                "time" => "%d minutes" % @parking.minutes,
                "paid" => @parking.paid, 
                "left" => @parking.left }, 200)
        else
            json_response({"message" => "payment required"}, 402)
        end
    end

    # GET /parking/:plate
    def historic

        if not params[:plate].match(/^[A-Z]{3}\-[0-9]{4}$/)
            return json_response({"message" => "invalid plate format"}, 400)
        end
        
        results = []
        Parking.where(plate: params[:plate]).find_each do |parking|
            if parking.minutes
                minutes = parking.minutes
            else
                diff_time = Time.now - parking.car_in
                minutes = (diff_time/60).to_i
            end
            results.push({
                "id" => parking.id,
                "time" => "%d minutes" % minutes,
                "paid" => parking.paid, 
                "left" => parking.left
            })
        end

        if results != []
            json_response(results, 200)
        else
            json_response({"message" => "plate not found"}, 404)
        end
    end

    private

    def parking_params
        # whitelist params
        params.permit(:plate)
    end

    def set_parking
        @parking = Parking.find(params[:id])
    end

    def set_parking_by_plate
        @parking = Parking.find(params[:plate])
    end

    def set_parkings_by_plate
        @parkings = Parking.find_by plate:(params[:plate])
    end

end
